packages/smithy-client/src/date-utils.ts

```diff
@@ -31,5 +34,267 @@ export function dateToUtcString(date: Date): string {
+ *
+ * @param value the value to parse
+ * @return a Date or undefined
+ */
+export const parseRfc7231DateTime = (value: unknown): Date | undefined => {
+  if (value === null || value === undefined) {
+    return undefined;
+  }
+  if (typeof value !== "string") {
+    throw new TypeError("RFC-7231 date-times must be expressed as strings");
+  }
+
+  //allow customization of day parsing for asctime days, which can be left-padded with spaces
+  let dayFn: (value: string) => number = (value) => parseDateValue(value, "day", 1, 31);
+
+  //all formats other than RFC 850 use a four-digit year
+  let yearFn: (value: string) => number = (value: string) => strictParseShort(value)!;
+  //RFC 850 dates need post-processing to adjust year values if they are too far in the future
+  let dateAdjustmentFn: (value: Date) => Date = (value) => value;
+
+  let match = IMF_FIXDATE.exec(value);
+  if (!match || !match.groups) {
+    match = RFC_850_DATE.exec(value);
+    if (match && match.groups) {
+      // RFC 850 dates use 2-digit years. So we parse the year specifically,
+      // and then once we've constructed the entire date, we adjust it if the resultant date
+      // is too far in the future.
+      yearFn = parseTwoDigitYear;
+      dateAdjustmentFn = adjustRfc850Year;
+    } else {
+      match = ASC_TIME.exec(value);
+      if (match && match.groups) {
+        dayFn = (value) => parseDateValue(value.trimLeft(), "day", 1, 31);
+      } else {
+        throw new TypeError("Invalid RFC-7231 date-time value");
+      }
+    }
+  }
+
+  const year = yearFn(match.groups["Y"]);
+  const month = parseMonthByShortName(match.groups["M"]);
+  const day = dayFn(match.groups["D"]);
+  return dateAdjustmentFn(buildDate(year, month, day, match));
+};
+
+const VALID_NUMBER = new RegExp(/^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$/);
+
+/**
+ * Parses a value into a Date. Returns undefined if the input is null or
+ * undefined, throws an error if the input is not a number or a parseable string.
+ *
+ * Input strings must be an integer or floating point number. Fractional seconds are supported.
+ *
+ * @param value the value to parse
+ * @return a Date or undefined
+ */
+export const parseEpochTimestamp = (value: unknown): Date | undefined => {
+  if (value === null || value === undefined) {
+    return undefined;
+  }
+
+  let valueAsDouble: number;
+  if (typeof value === "number") {
+    valueAsDouble = value;
+  } else if (typeof value === "string") {
+    // TODO: remove when https://github.com/aws/aws-sdk-js-v3/pull/2729 is merged, maybe
+    if (!VALID_NUMBER.exec(value)) {
+      throw new TypeError("Epoch timestamps must be valid floating point numbers.");
+    }
+    valueAsDouble = strictParseDouble(value)!;
+  } else {
+    throw new TypeError("Epoch timestamps must be expressed as floating point numbers or their string representation");
+  }
+
+  if (isNaN(valueAsDouble) || valueAsDouble === Infinity || valueAsDouble === -Infinity) {
```

Comment:
`Number.isNaN` will reject anything that isn't a number type.

```suggestion
  if (Number.isNaN(valueAsDouble) || valueAsDouble === Infinity || valueAsDouble === -Infinity) {
```